packages/backend/src/helpers/actions/handle-failed-step-and-throw.ts

```diff
@@ -0,0 +1,193 @@
+  504,
+  429, // Some 429s may not have Retry-After
+]
+
+function handleHttpError(
+  executionError: HttpError,
+  errorDetails: HandleFailedStepAndThrowParams['errorDetails'],
+): never {
+  // Handle reasonable Retry-After responses.
+  const retryAfterMs = parseRetryAfterToMs(
+    executionError.response?.headers?.['retry-after'],
+  )
+
+  if (retryAfterMs !== null) {
+    if (retryAfterMs <= RETRY_AFTER_LIMIT_MS) {
+      throw new RetriableError({
+        error: errorDetails,
+        delayInMs: retryAfterMs,
+        delayType: 'step',
+      })
+    }
+    throw new UnrecoverableError(`Retry-After (${retryAfterMs}ms) is too long!`)
+  }
+
+  // Handle retriable status codes
+  if (RETRIABLE_STATUS_CODES.includes(executionError.response?.status)) {
+    throw new RetriableError({
+      error: errorDetails,
+      delayInMs: 'default',
+      delayType: 'step',
+    })
+  }
+
+  // Handle retriable errors (identified by message substring)
+  const errorString = JSON.stringify(get(errorDetails, 'details.error', ''))
+  for (const errorSubstring of RETRIABLE_ERROR_SUBSTRINGS) {
+    if (errorString.includes(errorSubstring)) {
+      throw new RetriableError({
+        error: errorDetails,
+        delayInMs: 'default',
+        delayType: 'step',
+      })
+    }
+  }
+
+  // All other HTTP errors are un-retriable
+  throw new UnrecoverableError(JSON.stringify(errorDetails))
+}
+
+interface HandleFailedStepAndThrowParams {
+  errorDetails: ExecutionStep['errorDetails']
+
+  // This is thrown from app.run, which _in theory_ can be anything.
+  executionError: unknown
+
+  context: {
+    isQueuePausable: boolean
+    span: Span
+    worker: WorkerPro<IActionJobData>
+    job: JobPro<IActionJobData>
+  }
+}
+
+export function handleFailedStepAndThrow(
+  params: HandleFailedStepAndThrowParams,
+): never {
+  let { executionError } = params
+  const { errorDetails, context } = params
+  const { span, job } = context
+
+  // Edge case as some actions throw StepError now, but others don't.
+  if (executionError instanceof StepError) {
+    executionError = executionError.cause
+  }
+
+  // Inspect the error / failure reason, and determine if we can retry the step.
+  try {
+    if (executionError instanceof RetriableError) {
+      return handleRetriableError(executionError, context)
+    }
+    if (executionError instanceof HttpError) {
+      return handleHttpError(executionError, errorDetails)
+    }
+
+    // All other errors cannot be retried.
+    throw new UnrecoverableError(JSON.stringify(errorDetails))
+  } catch (finalError) {
+    // Update span and execution status as necessary.
+    const isRetriable =
+      !(finalError instanceof UnrecoverableError) &&
+      // -1 is needed because BullMQ only increments attemptsMade _after_ the
+      // job processor finishes, but we're currently still inside the job
+      // processor.
+      job.attemptsMade < MAXIMUM_JOB_ATTEMPTS - 1
+
+    if (isRetriable) {
```

Comment:
This is hella clunky because it duplicates the "isRetriable" logic from our `onFailed` worker callback.

I want to explore [`tracer.inject`](https://datadoghq.dev/dd-trace-js/interfaces/export_.Tracer.html#inject) to allow us to modify the same span within the callback, but too risky to make this change now (already a lot of changes in this PR)